### PR TITLE
Port Strech/Spot (stage 1)

### PR DESCRIPTION
**Pull request by jimmytyyang**
## Motivation and Context

 Porting Stretch and Spot robot into habitat sim. Either stretch or spot robot will be used in language rearrangement task. In addition, this will support future sim2real projects.

Plan:
(1) This is the first stage of porting robots. The diff is in robot folder.
(2) The second stage will add custom control and sensors for both Stretch and Spot
(3) The third stage will add yaml file to control and train the robots. I will wait until Hydra migration is done. For draft version of this, please refer to https://github.com/facebookresearch/habitat-lab/pull/956#issue-1384306723

 Chris worked on the initial design of stretch robot: https://github.com/cpaxton/hab_stretch/tree/cpaxton/hab-support

## How Has This Been Tested

 test_robot_wrapper.py has been added the support of testing stretch/spot robots

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by jimmytyyang** [comment]
Video for stretch robot via interactive_play.py 


https://user-images.githubusercontent.com/55121504/196302880-14a365f6-df08-4dcc-9e44-fbd9991c01ef.mp4

**Comment by dhruvbatra** [comment]
Awesome! 

Minor note -- can we get better material/lighting on that Stretch mesh? It looks so "flat".

**Comment by jimmytyyang** [comment]
[WIP] adding some textual using asset from Noah Maestre
https://user-images.githubusercontent.com/55121504/196820853-4098951f-e77e-4f37-b660-c4c4f748600e.mp4

**Comment by dhruvbatra** [comment]
Nicely getting there!

File: habitat-lab/habitat/robots/robot_base.py [hunk 1]
```diff
@@ -145,14 +225,113 @@ def is_base_link(self, link_id: int) -> bool:
...
+        if (
+            has_attr_contact_test and self._sim.habitat_config.contact_test
+        ) and (
+            (has_attr_navmesh and not self._sim.habitat_config.nav_mesh)
+            or (not has_attr_navmesh)
+        ):
+            did_collide = self._sim.contact_test(self.sim_obj.object_id)
+            robot_id = self.sim_obj.object_id
+            end_pos = target_rigid_state.translation
+            # If there is a collision, we do projection
```
Comment (aclegg3) [inline habitat-lab/habitat/robots/robot_base.py:258]: I'm not sure this is tested enough yet. For example, there is no validation that the projected position is collision free. It could be that projection from initial collisions causes new collisions.

Maybe consider removing the projection code for now and introducing that as a separate PR with tests.

File: habitat-lab/habitat/robots/robot_interface.py [hunk 2]
```diff
@@ -38,3 +40,54 @@ def get_link_and_joint_names(self) -> str:
...
+        return self.joint_motors[joint][1].position_target
+
+    def _set_motor_pos(self, joint, ctrl):
+        self._validate_joint_idx(joint)
+        self.joint_motors[joint][1].position_target = ctrl
+        self.sim_obj.update_joint_motor(
+            self.joint_motors[joint][0], self.joint_motors[joint][1]
+        )
+
+    def _vector_to_plane_proj(self, u, n):
```
Comment (aclegg3) [inline habitat-lab/habitat/robots/robot_interface.py:73]: These functions should probably be moved to a util module like[ geometry_utils.py](https://github.com/facebookresearch/habitat-lab/blob/main/habitat-lab/habitat/utils/geometry_utils.py)

File: habitat-lab/habitat/robots/spot_robot.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (aclegg3) [inline habitat-lab/habitat/robots/spot_robot.py:1]:
```suggestion
# Copyright (c) Meta Platforms, Inc. and its affiliates.
```

File: habitat-lab/habitat/robots/spot_robot.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+from habitat.robots.mobile_manipulator import (
+    MobileManipulator,
+    RobotCameraParams,
+)
+from habitat_sim.utils.common import orthonormalize_rotation_shear
+
+
+# TODO: refactor this class to support spherical joints: multiple dofs per link and #dofs != #positions
+@attr.s(auto_attribs=True, slots=True)
+class MobileManipulatorParams:
```
Comment (aclegg3) [inline habitat-lab/habitat/robots/spot_robot.py:20]: Looks like accidentally copied code

File: habitat-lab/habitat/robots/stretch_robot.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (aclegg3) [inline habitat-lab/habitat/robots/stretch_robot.py:1]:
```suggestion
# Copyright (c) Meta Platforms, Inc. and its affiliates.
```

**Review by aclegg3 — COMMENTED** [review]
Generally looks like a good first PR to introduce these robots. :+1: 

I've suggested a few refactors. Also would be ideal to get the stretch robot assets onto fbaipublicfiles and into the CI config for testing verification before we merge.

File: habitat-lab/habitat/robots/robot_base.py (hunk 1 above)
Comment (jimmytyyang) [reply habitat-lab/habitat/robots/robot_base.py]: Remove the projection code here. thank you!

File: habitat-lab/habitat/robots/robot_interface.py (hunk 2 above)
Comment (jimmytyyang) [reply habitat-lab/habitat/robots/robot_interface.py]: Remove the projection code. Thank you!

File: habitat-lab/habitat/robots/mobile_manipulator.py [hunk 6]
```diff
@@ -26,8 +26,9 @@ class RobotCameraParams:
     """
 
     attached_link_id: int
-    cam_offset_pos: mn.Vector3
-    cam_look_at_pos: mn.Vector3
+    cam_offset_pos: mn.Vector3 = mn.Vector3(0, 0, 0)
+    cam_look_at_pos: mn.Vector3 = mn.Vector3(0, 0, 0)
+    cam_orientation: mn.Vector3 = mn.Vector3(0, 0, 0)
```
Comment (Skylion007) [inline habitat-lab/habitat/robots/mobile_manipulator.py:31]:
```suggestion
    cam_offset_pos: mn.Vector3 = mn.Vector3.zero_init()
    cam_look_at_pos: mn.Vector3 = mn.Vector3.zero_init()
    cam_orientation: mn.Vector3 = mn.Vector3.zero_init()
```
I believe this is marginally faster.

Comment (jimmytyyang) [reply habitat-lab/habitat/robots/mobile_manipulator.py]: Thank you! :)

File: .circleci/config.yml [hunk 7]
```diff
@@ -149,7 +150,7 @@ jobs:
                 . activate habitat
                 pip install -U pip
                 conda install -q -y -c conda-forge ninja ccache numpy pytest pytest-mock pytest-cov psutil
-                pip install pytest-sugar mock cython pygame
+                pip install pytest-sugar>=0.9.6 mock cython pygame
```
Comment (aclegg3) [inline .circleci/config.yml:153]: Is this change necessary? Just want to confirm it has purpose.

**Review by aclegg3 — APPROVED** [review]
Looks like a solid step forward, let's merge it. :+1: 

Do need to rebase on newest changes from main and one small question about the ci config.
